json.go

```diff
@@ -0,0 +1,797 @@
+func (j JSONArray) GetElement(i int) Data {
+	n := int(j.dpiJsonArray.numElements)
+	elts := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonArray.elements)))[:n:n]
+	var d Data
+	jsonNodeToData(&d, &elts[i])
+	return d
+
+}
+
+func (j JSONArray) Get(nodes []Data) []Data {
+	n := int(j.dpiJsonArray.numElements)
+	elts := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonArray.elements)))[:n:n]
+	for i := 0; i < n; i++ {
+		var d Data
+		jsonNodeToData(&d, &elts[i])
+		nodes = append(nodes, d)
+	}
+	return nodes
+}
+
+func getJSONScalarDouble(d Data) (val float64, err error) {
+	err = nil
+	if d.NativeTypeNum == C.DPI_NATIVE_TYPE_BYTES {
+		b := (*C.dpiBytes)(unsafe.Pointer(&(d.dpiData).value))
+		s := C.GoStringN(b.ptr, C.int(b.length))
+		val, err = strconv.ParseFloat(s, 64)
+	} else {
+		val = (d.Get()).(float64)
+	}
+	return
+}
+
+func getJSONScalarString(d Data) (string, error) {
+	b := d.Get()
+	return string(b.([]byte)), nil
+}
+
+// Returns the Go type, []interface{} from JSONArray
+func (j JSONArray) GetValue() (nodes []interface{}, err error) {
+	n := int(j.dpiJsonArray.numElements)
+	elts := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonArray.elements)))[:n:n]
+	for i := 0; i < n; i++ {
+		var d Data
+		jsonNodeToData(&d, &elts[i])
+		if d.NativeTypeNum == C.DPI_NATIVE_TYPE_JSON_OBJECT {
+
+			jsobj := JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(d.dpiData))}
+			m, err := jsobj.GetValue()
+			if err != nil {
+				return nil, err
+			}
+			nodes = append(nodes, m)
+		} else if d.NativeTypeNum == C.DPI_NATIVE_TYPE_JSON_ARRAY {
+			jsarr := JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData))}
+			ua, err := jsarr.GetValue()
+			if err != nil {
+				return nil, err
+			}
+			nodes = append(nodes, ua)
+		} else {
+			if elts[i].oracleTypeNum == C.DPI_ORACLE_TYPE_VARCHAR {
+				keyval, err := getJSONScalarString(d)
+				if err == nil {
+					nodes = append(nodes, keyval)
+				} else {
+					return nil, err
+				}
+			} else if elts[i].oracleTypeNum == C.DPI_ORACLE_TYPE_NUMBER {
+				jd, err := getJSONScalarDouble(d)
+				if err != nil {
+					return nil, err
+				}
+				nodes = append(nodes, jd)
+			} else {
+				nodes = append(nodes, d.Get())
+			}
+		}
+	}
+	return nodes, nil
+}
+
+// Frees the C memory allocated
+func (jsarr *JSONArray) Close() error {
+	if jsarr.isMemOwned {
+		C.godror_dpiJsonfreeMem(jsarr.dpiJsonNode)
+	}
+	return nil
+}
+
+// It represents the map input.
+type JSONObject struct {
+	dpiJsonNode   *C.dpiJsonNode
+	dpiJsonObject *C.dpiJsonObject
+	isMemOwned    bool
+}
+
+func (j JSONObject) getOracleJSONType() C.dpiOracleTypeNum {
+	return C.DPI_ORACLE_TYPE_JSON_OBJECT
+}
+
+// populates dpiJsonNode from user inputs.
+// It creates a seperate memory for the new output value, jsonnode.
+// memory from user input, in is not shared with jsonnode.
+// Caller has to explicitly free using godror_dpiJsonfreeMem
+func populateJsonNode(in interface{}, jsonnode *C.dpiJsonNode) error {
```

Comment:
Make the first argument be the destination: `jsonnode *C.dpiJsonNode, in interface{}`.

Reply:
Fixed